fetch/peers/peers.go

```diff
@@ -151,12 +153,44 @@ func (p *Peers) SelectBestFrom(peers []peer.ID) peer.ID {
 func (p *Peers) SelectBest(n int) []peer.ID {
 	p.mu.Lock()
 	defer p.mu.Unlock()
+	return p.selectBest(n, nil)
+}
+
+// SelectBestWithProtocols is similar to SelectBest but filters peers by supported protocols.
+// If protocols is empty, it returns the best peers regardless of the protocol.
+// If protocols is not empty, it returns the best peers that support at least one of the protocols.
+func (p *Peers) SelectBestWithProtocols(n int, protocols []protocol.ID) []peer.ID {
```

Comment:
I'm not fully grasping this - why only at least one protocol needs to be supported? The only place outside of tests where this is used `activeSetProtocol` and `hashProtocol` are requested, aren't both needed?

Reply:
The idea is that there may be multiple versions of `syncv2` protocol deployed at some point in time: `sync/2`, `sync/2.1` and so on. When choosing peers for multipeer sync, any of these nodes will do, it's just that different protocols will be in use, later versions being more traffic-efficient etc.

For the purpose of hash fetching, the requirement may be just one of `activeSetProtocol` / `hashProtocol` or all of them, depending on which IDs are being fetched. It maybe could make some sense to invent a more complex logic for paritioning the set of hashes to download by the protocol etc., but in practice the servers for `activeSetProtocol` and `hashProtocol` are enabled almost at the same instant for most peers, and in a very unlikely case of a race the worst thing that may happen is current `go-spacemesh` behavior, that is, failed fetch which will be retried later, with failed peer getting lower priority when doing peer selection. So, in case of hash fetching, from practical standpoint it doesn't really matter whether we pass `activeSetProtocol`, `hashProtocol` or both to this function.